pkg/controller/manager/nad/controller.go

```diff
@@ -0,0 +1,334 @@
+	ctlbatchv1 "github.com/rancher/wrangler/pkg/generated/controllers/batch/v1"
+	"k8s.io/klog"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/harvester/harvester-network-controller/pkg/config"
+	"github.com/harvester/harvester-network-controller/pkg/utils"
+)
+
+const (
+	ControllerName = "harvester-network-manager-nad-controller"
+
+	jobContainerName      = "network-helper"
+	jobServiceAccountName = "harvester-network-helper"
+	JobEnvName            = "NADNETWORKS"
+
+	defaultInterface = "net1"
+
+	defaultPingTimes   = 5
+	defaultCheckPeriod = 15 * time.Minute
+)
+
+type nameWithNamespace struct {
+	namespace string
+	name      string
+}
+
+type checkMap struct {
+	items map[nameWithNamespace]string
+	mutex *sync.RWMutex
+}
+
+type Handler struct {
+	namespace   string
+	helperImage string
+
+	jobClient ctlbatchv1.JobClient
+	jobCache  ctlbatchv1.JobCache
+	nadClient ctlcniv1.NetworkAttachmentDefinitionClient
+	nadCache  ctlcniv1.NetworkAttachmentDefinitionCache
+
+	*checkMap
+}
+
+func Register(ctx context.Context, management *config.Management) error {
+	job := management.BatchFactory.Batch().V1().Job()
+	nad := management.CniFactory.K8s().V1().NetworkAttachmentDefinition()
+
+	handler := &Handler{
+		namespace:   management.Options.Namespace,
+		helperImage: management.Options.HelperImage,
+		jobClient:   job,
+		jobCache:    job.Cache(),
+		nadClient:   nad,
+		nadCache:    nad.Cache(),
+		checkMap: &checkMap{
+			items: make(map[nameWithNamespace]string),
+			mutex: new(sync.RWMutex),
+		},
+	}
+
+	go handler.CheckConnectivity()
+
+	nad.OnChange(ctx, ControllerName, handler.OnChange)
+	nad.OnRemove(ctx, ControllerName, handler.OnRemove)
+
+	return nil
+}
+
+func (h Handler) OnChange(key string, nad *cniv1.NetworkAttachmentDefinition) (*cniv1.NetworkAttachmentDefinition, error) {
+	if nad == nil || nad.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	// check annotations
+	if nad.Annotations == nil || nad.Annotations[utils.KeyNetworkConf] == "" {
+		return nad, nil
+	}
+	networkConf, err := utils.NewLayer3NetworkConf(nad.Annotations[utils.KeyNetworkConf])
+	if err != nil {
+		return nil, fmt.Errorf("invalid layer 3 network configure: %w", err)
+	}
+
+	if networkConf.CIDR != "" && networkConf.Gateway != "" {
+		// set connectivity as the initial status unknown
+		if networkConf.Connectivity == "" {
+			if err := h.setUnknown(nad, networkConf); err != nil {
+				return nil, err
+			}
+		}
+		// add item to map
+		h.addItem(nad.Namespace, nad.Name, networkConf.Gateway)
+		return nad, nil
+	}
+
+	if networkConf.Mode != utils.Auto {
+		return nad, nil
+	}
+	// create or update job to get layer 3 network information automatically
+	if err := h.createOrUpdateJob(nad); err != nil {
+		return nil, err
+	}
```

Comment:
need to sync the configured status from the job pod to network CR.

Reply:
We don't need to sync the status because if the job runs successfully, the connectivity will be set as `unknown` in the annotation of nad CR, otherwise, it will be empty.